### PR TITLE
New domain: SWE

### PR DESCRIPTION
**Pull request by ehsk**
RL on SWE-style tasks like SWE-smith or SWE-bench added.

Experimented with Qwen2.5-7B-Instruct (grey) and Qwen3-8B (orange).

image

Credit to: @amilios (this code is mostly taken from his implementation)

### CODE REVIEW

**Review by amilios — APPROVED** [review]
lgtm

File: pipelinerl/finetune/value_model.py [hunk 1]
```diff
@@ -107,7 +106,6 @@ def forward(
         values = self.value_head(hidden_states)
 
         return CausalLMOutputWithValue(
-            loss=outputs.loss,
             logits=outputs.logits,
```
Comment (rafapi) [inline pipelinerl/finetune/value_model.py:105]: value-head model will fail now, we should revert this

File: conf/swe.yaml [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+dataset_loader_params:
+  seed: ${seed}
+  # max_samples: 1000  # uncomment to cap the number of loaded samples (applies to both train and test)
+
+# HuggingFace Hub dataset IDs (or local disk paths).
+# Append ":split" to restrict to a specific split, e.g. SWE-bench/SWE-smith-py:train
+train_dataset_names:
+  - SWE-bench/SWE-smith-py
+test_dataset_names:
+  - SWE-bench/SWE-smith-py
```
Comment (rafapi) [inline conf/swe.yaml]: same data for train and test? we may not be running tests by default but we should not have this as a default just in case we do.

File: pipelinerl/domains/swe/reward.py [hunk 3]
```diff
@@ -0,0 +1,163 @@
+import difflib
+import logging
+import re
+from typing import Dict, List, Tuple, TypedDict
+
+from unidiff import PatchSet
+from unidiff.errors import UnidiffParseError
```
Comment (rafapi) [inline pipelinerl/domains/swe/reward.py:7]: is unidiff in pyporject.toml?

**Review by rafapi — CHANGES_REQUESTED** [review]
we should also add the this new domain to the multi-domain integration layer:

- pipelinerl/domains/multidomain/loader.py
- conf/multi_domain/base.yaml
- conf/domain_rollouts/base.yaml

File: pipelinerl/finetune/value_model.py (hunk 1 above)
Comment (ehsk) [reply pipelinerl/finetune/value_model.py:105]: The loss is computed outside, we only need to return the value head tensors: https://github.com/ServiceNow/PipelineRL/blob/swe/pipelinerl/finetune/rl/__init__.py#L262

File: pipelinerl/domains/swe/reward.py (hunk 3 above)
Comment (ehsk) [reply pipelinerl/domains/swe/reward.py:7]: Yes, added as an optional dependency similar to other domains

File: conf/swe.yaml (hunk 2 above)
Comment (ehsk) [reply conf/swe.yaml]: Good catch! updated the default datasets.

File: pipelinerl/finetune/value_model.py (hunk 1 above)
Comment (rafapi) [reply pipelinerl/finetune/value_model.py:105]: I see, that should be fine, not sure it's the most flexible way to use a value head though. But anyway, with this change i see a failure path in `pipelinerl/finetune/eval.py`. Also, you should remove the loss from `CausalLMOutputWithValue` and labels from `forward()`

Comment (ehsk) [reply pipelinerl/finetune/value_model.py:105]: Sure, removed the extra args: `labels` in `forward` and `loss` in `CausalLMOutputWithValue`

About `finetune/eval.py`, in theory, yes but eval.py is an old code that's never used anywhere else.
